File: algorithm/_misc/merge-k-sorted-arrays.cpp
// Merge K Sorted Arrays

vector<int> mergeKArrays(vector<vector<int>>sorted) {
  vector<int> result;
  pririty_queue<int, vector<int>, less<int>> minHeap;

  for (int i = 0; i < sorted.size(); i++) {
    if (!sorted[i].empty()) {
      minHeap.push(make_tuple(sorted[i][0], i, 0));
    }
  }
  
  while (!minHeap.empty()) {
    int e, i, j;
    tie(e, i , j) = minHeap.top(); minHeap.pop();

    result.push_back(e);
    
    if (j + 1 < sorted[i].size()) {
      minHeap.push(make_tuple(sorted[i][j + 1], i, j + 1));
    }
  }
  return result;
}

File: data-structure-and-algorithm/typescript/leetcode/src/extra-questions/merge-k-sorted-arrays.ts
/*
Merge K Sorted Arrays
*/

import {Heap} from '../data-structure/Heap';

function mergeKArrays(sortedCollection: number[][]) {
  const heap = new Heap<[number[], number]>(([list1, idx1], [list2, idx2]) => {
    return list1[idx1] - list2[idx2];
  });

  for (const list of sortedCollection) {
    if (list.length >= 1) {
      heap.push([list, 0]);
    }
  }

  const result: number[] = [];
  while (!heap.isEmpty()) {
    const [list, idx] = heap.pop();
    result.push(list[idx]);

    if (idx + 1 < list.length) {
      heap.push([list, idx + 1]);
    }
  }
  return result;
}

console.log(
  mergeKArrays([
    [1, 3, 5, 7, 9],
    [2, 4, 6, 8, 10],
    [0, 12],
  ])
);
